lvgl/src/display.rs

```diff
@@ -149,67 +150,56 @@ impl DefaultDisplay {
 /// entire number of pixels on the screen, in which case the screen will be
 /// drawn to multiple times per frame.
 pub struct DrawBuffer<const N: usize> {
-    //inner: NonNull<lvgl_sys::lv_disp_draw_buf_t>,
-    initialized: RunOnce,
-    refresh_buffer: RefCell<[MaybeUninit<lvgl_sys::lv_color_t>; N]>,
+    draw_buf: lvgl_sys::lv_disp_draw_buf_t,
+    _refresh_buffer: Pin<Box<[MaybeUninit<lvgl_sys::lv_color_t>; N]>>,
+    //_refresh_buffer: [MaybeUninit<lvgl_sys::lv_color_t>; N],
 }
 
 impl<const N: usize> Default for DrawBuffer<N> {
     fn default() -> Self {
+        let mut buf = Box::pin([MaybeUninit::uninit(); N]);
         Self {
-            initialized: RunOnce::new(),
-            refresh_buffer: RefCell::new([MaybeUninit::uninit(); N]),
-        }
-    }
-}
-
-impl<const N: usize> DrawBuffer<N> {
-    fn get_ptr(&self) -> Option<Box<lvgl_sys::lv_disp_draw_buf_t>> {
-        if self.initialized.swap_and_check() {
-            // TODO: needs to be 'static somehow
-            // Cannot be in the DrawBuffer struct because the type `lv_disp_buf_t` contains a raw
-            // pointer and raw pointers are not Send and consequently cannot be in `static` variables.
-            let mut inner: MaybeUninit<lvgl_sys::lv_disp_draw_buf_t> = MaybeUninit::uninit();
-            let primary_buffer_guard = &self.refresh_buffer;
-            let draw_buf = unsafe {
+            draw_buf: unsafe {
+                let mut inner: MaybeUninit<lvgl_sys::lv_disp_draw_buf_t> = MaybeUninit::uninit();
+                let raw_ptr = buf.as_mut_ptr() as *mut _;
                 lvgl_sys::lv_disp_draw_buf_init(
                     inner.as_mut_ptr(),
-                    primary_buffer_guard.borrow_mut().as_mut_ptr() as *mut _,
+                    raw_ptr,
                     ptr::null_mut(),
                     N as u32,
                 );
                 inner.assume_init()
-            };
-            Some(Box::new(draw_buf))
-        } else {
-            None
+            },
+            _refresh_buffer: buf,
         }
     }
 }
 
+impl<const N: usize> DrawBuffer<N> {
+    fn get_ptr(&mut self) -> &mut lvgl_sys::lv_disp_draw_buf_t {
+        &mut self.draw_buf
+    }
+}
+
 #[repr(C)]
 pub(crate) struct DisplayDriver<const N: usize> {
-    pub(crate) disp_drv: lvgl_sys::lv_disp_drv_t,
+    pub(crate) disp_drv: Box<lvgl_sys::lv_disp_drv_t>,
     _buffer: DrawBuffer<N>,
 }
 
 impl<'a, const N: usize> DisplayDriver<N> {
-    pub fn new<F>(draw_buffer: DrawBuffer<N>, display_update_callback: F) -> Result<Self>
+    pub fn new<F>(mut draw_buffer: DrawBuffer<N>, display_update_callback: F) -> Result<ManuallyDrop<Self>>
```

Comment:
This makes it so the actual driver doesn't get dropped anymore; not happy about it, but the alternative would be propagating the genericness of `DrawBuffer` all over the place